### PR TITLE
RFC6121 rosterver, XEP-0352 CSI and stream features extraction

### PR DESCRIPTION
**Pull request by rufferson**
This set of 3 patches 
1) adds ability to extract stream features - since not all of them related to wocky space.
In particular it extracts CSI, AMP, rosterver and SM features flagging them on connection feature mask.
2) extends google:queue functionality to support client state indication (CSI) as per XEP-0352
3) adds implementation of the roster versioning and consequently roster cache. Roster is cached in sqlite3 db and cache is serialized and injected as WockyNode tree  into roster processing.

### CODE REVIEW

**Comment by Kaffeine** [comment]
Hi there. I can't review non-trivial patches and it would be unresponsible to just merge PRs.
@rufferson and @noonien-d You both seem to know the code and it is a real chance to move forward. I think it would be acceptable if you'll review each-other PRs (and I merge them).

@noonien-d can you please review this series?

**Comment by rufferson** [comment]
Rebased on master's tip, fixed styling (to pass make check) and added some make-up to make it build under GCC9 (Arch). Note - still doesn't include new wocky api

**Comment by noonien-d** [comment]
I'll review the next days.
@rufferson would you mind to add an additional PR to enable the wocky api changes?

**Comment by rufferson** [comment]
I'd expect it to be bundled with commit updating the submodule pointer, otherwise the change would breaking and will need another commit to recover. I was thinking today adding conditional interface change but even that is non reliable as commit which removes peername from callback does not change any macro definitions.

**Comment by Kaffeine** [comment]
Do you have a version definition in wocky? Like `OPENSSL_VERSION_NUMBER`?
It is a typical mean to introduce changes (especially backward incompatible).
We had to do the same in telepathy-nonsense: https://github.com/TelepathyIM/telepathy-nonsense/blob/master/connection.cc#L242

**Comment by rufferson** [comment]
Unfortunatelly no, it was carved out from gabble without that, currently version is statically defined in wocky/wocky.pc as 0.0.0.1

File: configure.ac [hunk 1]
```diff
@@ -220,7 +221,7 @@ PKG_CHECK_MODULES(GLIB,
 PKG_CHECK_MODULES(GMODULE, [gmodule-2.0 >= 2.32])
 
 AC_DEFINE(GLIB_VERSION_MIN_REQUIRED, GLIB_VERSION_2_32, [Ignore post 2.32 deprecations])
-AC_DEFINE(GLIB_VERSION_MAX_ALLOWED, GLIB_VERSION_2_32, [Prevent post 2.32 APIs])
+AC_DEFINE(GLIB_VERSION_MAX_ALLOWED, GLIB_VERSION_2_34, [Prevent post 2.32 APIs])
```
Comment (noonien-d) [inline configure.ac:224]: Minor: description should be "Prevent post 2.32 APIs"

File: src/capabilities.c [hunk 2]
```diff
@@ -96,10 +96,12 @@ static const Feature self_advertised_features[] =
   { 0, NULL }
 };
 
+#if 0
```
Comment (noonien-d) [inline src/capabilities.c:99]: Remove this if its not needed?

File: src/roster-cache.c [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+}
+
+static gchar *
+get_path (void)
+{
+  gchar *free_dir = NULL, *ret;
+  const gchar *dir, *path;
+
+  /* This should still work if it's the magic value ":memory:". */
+  path = g_getenv ("ROSTER_CACHE");
```
Comment (noonien-d) [inline src/roster-cache.c:152]: where does `ROSTER_CACHE` come from?

File: src/roster.c [hunk 4]
```diff
@@ -1826,13 +1870,28 @@ connection_status_changed_cb (GabbleConnection *conn,
...
+                  user = conn_util_get_bare_self_jid (conn);
+                  if (user)
+                      query = roster_cache_get_roster (self->priv->rcache, user);
+                  if (query)
+                    {
+                      process_roster (self, query);
+                      wocky_node_free (query);
+                    }
+                }
               stanza = _gabble_roster_message_new (self, WOCKY_STANZA_SUB_TYPE_GET,
```
Comment (noonien-d) [inline src/roster.c:1895]: Shouldn't you avoid requesting the roster after you loaded the cache successfully? Do I miss something?

Comment (rufferson) [reply src/roster.c:1895]: No, we need to request roster updates since last cached version. Cache allows us to save the bandwidth as without any changes we'd get simple <iq type=result/> as confirmation. Otherwise we have two possibilities - server floods updates at us or send full roster (up to server to decide what's better)

Comment (noonien-d) [reply src/roster.c:1895]: I see, thanks for the explanation.

File: src/roster.c [hunk 5]
```diff
@@ -1410,7 +1445,13 @@ got_roster_iq (GabbleRoster *roster,
       return FALSE;
     }
 
-  process_roster (roster, query_node);
+  if (query_node)
+    {
+      process_roster (roster, query_node);
+      user = conn_util_get_bare_self_jid (priv->conn);
+      DEBUG ("updating roster cache for %s", user);
+      roster_cache_update_roster (roster->priv->rcache, user, query_node);
```
Comment (noonien-d) [inline src/roster.c:1453]: Needs to check if rcache != NULL, in case the server doesn't support rosterver.

File: src/roster-cache.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+GType
+roster_cache_get_type (void);
+
+WockyNode *roster_cache_get_roster (RosterCache *self,
+    const gchar *user);
+
+gboolean roster_cache_update_roster (RosterCache *,
+    const gchar *, WockyNode *);
+
+RosterCache *
```
Comment (noonien-d) [inline src/roster-cache.h:71]: No need to expose roster_cache_new

File: plugins/console/plugin.c [hunk 7]
```diff
@@ -76,6 +76,7 @@ plugin_iface_init (
 
   iface->name = "XMPP console";
   iface->version = PACKAGE_VERSION;
+  iface->sidecar_interfaces = sidecar_interfaces;
```
Comment (noonien-d) [inline plugins/console/plugin.c:79]: I'm not sure about this, why did you add it?

Comment (rufferson) [reply plugins/console/plugin.c:79]: this eventually a fix for this sidecar plugin as the reason the static variable was unused is because this line was missing. The sidecar though is rather example so no one ever cared I presume whether it actually worked. I can probably move it to separate commit as there's also related python script (client side of the sidecar interface - to test its functionality) which also needs similar one line fix to make it work and the commit title doesn't match the purpose

**Comment by rufferson** [comment]
Thanks for review, that was really helpful. 
I cannot think of any sensible way to handle it except to schedule the retry for later which seems overly complicated. Even if cache is stuck - it's non-breaking issue, each login server will be sending the diff, after the diff grows too big server will simply start sending entire roster (ignoring the cache). What I will do perhaps is just make it visible (Err-out, currently it's debug from cache itself) so that someone can take a look and either fix or drop it.

**Comment by rufferson** [comment]
btw I see you have implemented CSI in wocky, why didn't you flag it here? I think it makes more sense having it in wocky on top of SM resumption queue. And I've pushed it here just because back then I considered wocky to be overly complex :) I'll gladly ditch it in favour of reviving SM/CSI wocky branch

**Comment by noonien-d** [comment]
Actually I wondered if its not nicer to keep in in gabble, if possible. Can you explain why there is the need to keep it with sm?

**Comment by rufferson** [comment]
My line of thinking is that on mobile devices we want to minimize radio usage, so if we got powersave indication and flushed the queue - we do not want to do any residual writes anymore to avoid waking/inhibiting sleep of the radio accidentally. Therefore after we sent inactive we don't write but queue instead, and on active - flush again.

**Comment by noonien-d** [comment]
Why don't we enable power saving mode in gabble when google queue is available?

**Comment by rufferson** [comment]
well... yes, actually one (csi) could be independent from another (wocky queue)... so we may make it configurable to enforce queue in addition to indication for mobile platforms. Ok, convinced :)

**Comment by Kaffeine** [comment]
@rufferson I'm ready to merge this once you confirm it.

**Comment by rufferson** [comment]
was thinking to add gcc10 cleanups here but will probably rather push them with next pr for carbons management. So yes, good to go. Thanks for review.
